### PR TITLE
fix #244, add more UT

### PR DESCRIPTION
**Pull request by clockfly**
[Review on Reviewable](https://reviewable.io/reviews/intel-hadoop/gearpump/255)

### CODE REVIEW

**Comment by clockfly** [comment]
Only need to review https://github.com/clockfly/gearpump/commit/473416932b0865225663fa87867c5ab053216ddf
in this PR, as it depends on some old PR.

File: core/src/main/resources/reference.conf [hunk 1]
```diff
@@ -1,128 +1,141 @@
...
-    scheduling {
-      scheduler-class = "org.apache.gearpump.cluster.scheduler.PriorityScheduler"
-    }
+  ### This path follows the hadoop path schema
+  ### For HDFS, use hdfs://host:port/path/
+  ### For FTP, use ftp://host:port/path
+  ### If you want to store on master nodes, then use local directory,
+  ### jarstore.rootpath = "jarstore/" will points to relative directory where master is started.
+  ### jarstore.rootpath = "/jarstore/" will points to absolute directory on master server
+  jarstore.rootpath = "jarstore/"
```
Comment (huafengw) [inline core/src/main/resources/reference.conf]: When we pack-archive gearpump and unzip the tar file, there is only gear.conf under the conf folder, so ```jarstore.rootpath``` is not configurable for a user, is it acceptable?

File: core/src/main/scala/org/apache/gearpump/cluster/Application.scala [hunk 2]
```diff
@@ -29,8 +29,53 @@ case class AppJar(name: String, container: JarFileContainer)
...
+/**
+ * This interface should keep stable, so that we can remain
+ * compatible with old versions
+ */
+trait ExecutorContextInterface {
+  val executorId : Int
+  val workerId : Int
+  val appId : Int
+  val appMaster : ActorRef
+  val startClock : TimeStamp
```
Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/cluster/Application.scala]: I think startClock field is mainly related to the streaming type application, so is it necessary for this general interface?

File: core/src/main/scala/org/apache/gearpump/transport/netty/Context.scala [hunk 3]
```diff
@@ -63,9 +63,9 @@ import org.apache.gearpump.transport.netty.Context._
   }
 
 
-  def bind(name: String, lookupActor : ActorLookupById): Int = {
+  def bind(name: String, lookupActor : ActorLookupById, deserializeFlag : Boolean = true): Int = {
     val taskDispatcher = system.settings.config.getString("gearpump.task-dispatcher")
```
Comment (huafengw) [inline core/src/main/scala/org/apache/gearpump/transport/netty/Context.scala]: Is this flag an application-level configuration? If so you can move it to the user configuration.

File: core/src/test/scala/org/apache/gearpump/cluster/TestUtil.scala [hunk 4]
```diff
@@ -53,4 +53,14 @@ object TestUtil{
 
     def shutDown() = system.shutdown()
   }
-}
+
+  def findFreePort: Try[Int] = {
```
Comment (huafengw) [inline core/src/test/scala/org/apache/gearpump/cluster/TestUtil.scala]: Duplicated function in core/src/main/scala/org/apache/gearpump/util/Util.scala

**Comment by whjiang** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/intel-hadoop/gearpump/255)







---

<sup>**[core/src/test/scala/org/apache/gearpump/cluster/main/MainSpec.scala, line 45 \[r2\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/255#-JesuCjsujypxkYg-f3s)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/a0e2f3ee2a7e8f0e4a271e3e6f7646a2bb883721/core/src/test/scala/org/apache/gearpump/cluster/main/MainSpec.scala#L45)):</sup>
it seems the comment is not consistent with the real code (2000ms vs 10s).

---

<sup>**[core/src/test/scala/org/apache/gearpump/cluster/main/MainSpec.scala, line 121 \[r2\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/255#-Jet17RPFXo3lOTCNv80)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/a0e2f3ee2a7e8f0e4a271e3e6f7646a2bb883721/core/src/test/scala/org/apache/gearpump/cluster/main/MainSpec.scala#L121)):</sup>
can we group the creation of mock master into a function to avoid duplication?

---

<sup>**[examples/fsio/src/main/scala/org/apache/gearpump/streaming/examples/fsio/SeqFileStreamProcessor.scala, line 50 \[r2\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/255#-Jet2NasO3PfkEGz4C-0)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/a0e2f3ee2a7e8f0e4a271e3e6f7646a2bb883721/examples/fsio/src/main/scala/org/apache/gearpump/streaming/examples/fsio/SeqFileStreamProcessor.scala#L50)):</sup>
better rename taskContext argument. It is quite confusing with the constructor param. 

maybe a better name is startTime. Why we name the type as NewStartTime? Does it mean that onStart is not called only once per task actor?

---

<sup>**[streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskContext.scala, line 24 \[r2\]](https://reviewable.io:443/reviews/intel-hadoop/gearpump/255#-Jet254jp_GJFx4nNh50)** ([raw file](https://github.com/intel-hadoop/gearpump/blob/a0e2f3ee2a7e8f0e4a271e3e6f7646a2bb883721/streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskContext.scala#L24)):</sup>
It's great we have TaskContext for each task now!!!!

---

File: core/src/main/resources/reference.conf (hunk 1 above)
Comment (clockfly) [inline core/src/main/resources/reference.conf]: There is also a section "jarstore.rootpath" in gear.conf
which will override reference.conf.

**Comment by clockfly** [comment]
@whjiang, good points, I have made the changes.

File: core/src/main/scala/org/apache/gearpump/transport/netty/Context.scala (hunk 3 above)
Comment (clockfly) [inline core/src/main/scala/org/apache/gearpump/transport/netty/Context.scala]: good point!

I now add a TODO comment in the source code.
I am not sure whether we want to allow a application to override this config for now.

File: core/src/main/scala/org/apache/gearpump/cluster/Application.scala (hunk 2 above)
Comment (clockfly) [inline core/src/main/scala/org/apache/gearpump/cluster/Application.scala]: good point.

**Comment by clockfly** [comment]
I have fixed all the comments. Please check 
https://github.com/clockfly/gearpump/commit/89f2d2299aa5a145d4e865777db30d5685028c48
 for changes.

**Comment by clockfly** [comment]
Please suggest if there is other concern.

**Comment by huafengw** [comment]
+1

**Pull request merged**